truffle/src/main/java/org/jruby/truffle/core/regexp/RegexpNodes.java

```diff
@@ -426,7 +430,25 @@ public Object matchGeneric(VirtualFrame frame, DynamicObject regexp, DynamicObje
                 toStrNode = insert(ToStrNodeGen.create(getContext(), getSourceSection(), null));
             }
 
-            return match(regexp, toStrNode.executeToStr(frame, other));
+            return matchWithStringCopy(regexp, toStrNode.executeToStr(frame, other));
+        }
+
+        // Creating a MatchData will store a copy of the source string. It's tempting to use a rope here, but a bit
+        // inconvenient because we can't work with ropes directly in Ruby (useful for implementing some MatchData
+        // methods). Likewise, we need to taint objects based on the source string's taint state. We mustn't allow the
+        // source string's contents to change, however, so we must ensure that we have a private copy of that string.
+        private Object matchWithStringCopy(DynamicObject regexp, DynamicObject string) {
+            final Matcher matcher = createMatcher(getContext(), regexp, string);
+            final int range = StringOperations.rope(string).byteLength();
+
+            final DynamicObject matchData = matchCommon(getContext(), this, makeSubstringNode, regexp, string, true, matcher, 0, range);
+            setLastMatchNode.executeSetLastMatch(matchData);
+
+            if (matchData != nil()) {
+                return matcher.getBegin();
+            }
```

Comment:
In the case of the String specialization, I continue to `dup` it. I guess strictly speaking some non-String object could cache its `to_s` output, but wouldn't that violate Ruby semantics? Even Symbols get newly allocated Strings from `to_s`.

We do need an explicit copy (or, switch to storing taint state and a rope and updating everything using the source string as necessary). The change was made to fix an issue whereby you can mutate the source string from outside the MatchData via the shared reference.